### PR TITLE
Reduce time for integration tests

### PR DESCRIPTION
**Pull request by Tulsishah**
Today all the tests are running sequentially, because of that we have to wait almost 1 hour to check e2e test success and it is affecting the acceleration of the development cycle.

### Packages in integration tests
Currently there are 14 packages in our integration test directory.

1. explicit_dir
2. implicit_dir
3. list_large_dir
4. operations
5. read_large_files
6. rename_dir_limit
7. managed_folders
8. read_only
9. local_file
10. log_rotation
11. mounting
12. read_cache
13. gzip
14. write_large_files

Packages 1-6: Must run one after the other. Each individual test within these packages also needs to run sequentially. This is because they clean a shared directory (mnt dir).
Packages 7-8 (readonly and managed_folders) Cannot be run in parallel at all. They alter permissions that would break other tests.
Packages 9-14: Can run in parallel with each other (package-level parallelization). However, tests within each package must run sequentially since they work on the same directory.

We can parallelize packages 1 to 6 slowly, it will require effort.
## Approach 
Basic Idea: We're splitting the tests into three groups to make things faster.

1. Three buckets: We're using separate spaces to avoid tests bumping into each other.
2. Group 1 (Non-parallel): These tests must run one after the other. To save time, they'll work quietly in the background. These tests can parallelize with some additional effort and we will do it gradually.
3. Group 2 (Never parallelize): These tests can't be parallelize ever as they are changing bucket permissions.
4. Group 3 (Parallel): These tests can all work at the same time, each handling their own package.

We removed the `setup.RemoveBinFileCopiedForTesting()` function because it caused conflicts between tests. One test would delete a file that another test needed. We addressed this with a script, but running integration tests could [reintroduce the file](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/tools/integration_tests/util/setup/setup.go#L187). Therefore, removing the function permanently shouldn't cause any problems.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - kokoro
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (raj-prince) [inline perfmetrics/scripts/run_e2e_tests.sh:56]: Instead pass take bucket_prefix as an argument or $1 and create bucket based on that.

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (raj-prince) [inline perfmetrics/scripts/run_e2e_tests.sh:109]: Call the create_bucket with bucketPrefix as an argument.

**Review by raj-prince — COMMENTED** [review]
Approach looks good to me, one minor comment.

Our next goal would be make all the non-parallel test package independent (other than two package which required bucket permission change).

**Review by sethiay — COMMENTED** [review]
Before making changes, could we ensure this doesn't cause failures in GKE pipeline ?

**Comment by Tulsishah** [comment]
> Before making changes, could we ensure this doesn't cause failures in GKE pipeline ?

These changes will not break GKE pipeline as we are not making any change for mountedDirectory flag. Changes are for kokoro e2e script only. Although I'll run that script which we are sharing with kokoro team before merging.

**Comment by Tulsishah** [comment]
> Approach looks good to me, one minor comment.
> 
> Our next goal would be make all the non-parallel test package independent (other than two package which required bucket permission change).

Yes I updated in document.

Fixed comments. PTAL

**Review by raj-prince — DISMISSED** [review]

**Comment by raj-prince** [comment]
Please wait for the @sethiay's approval before merge.

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (Tulsishah) [inline perfmetrics/scripts/run_e2e_tests.sh:85]: Created separate function as calling same function with argument in background was overriding the value.

FYI @raj-prince @sethiay

Comment (ankitaluthra1) [reply perfmetrics/scripts/run_e2e_tests.sh:85]: @Tulsishah is $test_fail_np_group_1 value getting overridden if we use same function ? if that's the case then can we append a prefix/suffix of bucket name in this variable name and keep the function same? will it help ?

Comment (Tulsishah) [reply perfmetrics/scripts/run_e2e_tests.sh:85]: `test_dir_np_group_1` the array itself is overriding, so all the tests are not executing that's what I face. It's not about failure code, it's about execution itself.

Comment (Tulsishah) [reply perfmetrics/scripts/run_e2e_tests.sh]: It is fixed. Thanks @ankitaluthra1 I was capturing arrays properly in functions that's why it was happening.

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:81]: should we print test_dir_np instead of test_fail_np?

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:132]: the create bucket is called after this echo statement

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:165]: where is test_fail_np used ?

**Review by sethiay — COMMENTED** [review]
Also, could you please confirm that the gcsfuse debug logs of different tests would not interfere with each ?

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (Tulsishah) [reply perfmetrics/scripts/run_e2e_tests.sh:165]: On line number 80

**Comment by Tulsishah** [comment]
> Also, could you please confirm that the gcsfuse debug logs of different tests would not interfere with each ?

They should not as every package is building gcsfuse in different directory `/tmp/gcsfuse_readwrite_test_xxx/`, so logfile will be differently stored in these package and in case of failure we are sending logs to kokoro bucket.

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:94]: correct this

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:96]: make all the local variables follow the convention lower_case

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:26]: you can make this CAPS_LOCK

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:173]: parallel_tests_pid

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:182]: parallel_tests_exit_code

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:141]: keep it at top and make it CAPS_LOCK

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:151]: nit: parallel

File: perfmetrics/scripts/run_e2e_tests.sh
Comment (sethiay) [inline perfmetrics/scripts/run_e2e_tests.sh:173]: nit: Add logical space

**Review by sethiay — APPROVED** [review]

**Review by sethiay — APPROVED** [review]
Merge after the Kokoro tests pass.

**Pull request merged**
